src/mds/MDBalancer.cc

```diff
@@ -99,7 +99,13 @@ void MDBalancer::handle_export_pins(void)
     auto cur = it++;
     CInode *in = *cur;
     ceph_assert(in->is_dir());
-    mds_rank_t export_pin = in->get_export_pin(false);
+    mds_rank_t export_pin = MDS_RANK_NONE;
+    // Making sure the ephemeral pin does not override export pin
+    if (in->get_export_pin(false) != MDS_RANK_NONE)
+      export_pin = in->get_export_pin(false);
+    else if (in->is_export_ephemeral_distributed_migrating || in->is_export_ephemeral_random_migrating)
```

Comment:
You're missing `{}` here which causes a bug: (non-ephemeral) export pins are always overridden.

Coding lesson: always use `{}` for control statements. The LOC saved is never worth it and you may hit this kind of coding error whenever you add another line.